NDTensors/src/blocksparse/blocksparse.jl

```diff
@@ -12,21 +14,29 @@ struct BlockSparse{ElT,VecT,N} <: TensorStorage{ElT}
   end
 end
 
-# TODO: Implement as `fieldtype(storagetype, :data)`.
-datatype(::Type{<:BlockSparse{<:Any,DataT}}) where {DataT} = DataT
-# TODO: Implement as `ndims(blockoffsetstype(storagetype))`.
-ndims(storagetype::Type{<:BlockSparse{<:Any,<:Any,N}}) where {N} = N
 # TODO: Implement as `fieldtype(storagetype, :blockoffsets)`.
 blockoffsetstype(storagetype::Type{<:BlockSparse}) = BlockOffsets{ndims(storagetype)}
 
-function set_datatype(storagetype::Type{<:BlockSparse}, datatype::Type{<:AbstractVector})
-  return BlockSparse{eltype(datatype),datatype,ndims(storagetype)}
+function TypeParameterAccessors.position(
+  ::Type{<:BlockSparse}, ::typeof(TypeParameterAccessors.parenttype)
+)
+  return Position(2)
+end
+TypeParameterAccessors.position(::Type{<:BlockSparse}, ::typeof(Base.ndims)) = Position(3)
+TypeParameterAccessors.parameter_name(::Type{<:BlockSparse}, ::Position{2}) = parenttype
+TypeParameterAccessors.parameter_name(::Type{<:BlockSparse}, ::Position{3}) = Base.ndims
+function TypeParameterAccessors.default_parameter(
+  type::Type{<:BlockSparse}, ::typeof(TypeParameterAccessors.parenttype)
+)
+  return Vector{default_parameter(type, eltype)}
 end
 
-function SetParameters.set_ndims(storagetype::Type{<:BlockSparse}, ndims::Int)
-  return BlockSparse{eltype(storagetype),datatype(storagetype),ndims}
+function TypeParameterAccessors.set_ndims(storagetype::Type{<:BlockSparse}, param::Int)
+  return set_parameter(storagetype, Base.ndims, param)
```

Comment:
```suggestion
  return set_parameter(storagetype, ndims, param)
```

Reply:
I think whats happening here is `NDTensors` has an `ndims` function nhttps://github.com/ITensor/ITensors.jl/blob/8e1b8f128244009280c8ad42e5c925d178f4c99e/NDTensors/src/abstractarray/ndims.jl#L3 
So I am wondering, should I reroute this function to use `TypeParameters` or should I migrate the code to use `Base.ndims` ?

Reply:
I made `NDTensors.ndims` call `TypeParameterAccessors.parameter(type, ndims)` and the code seems to work for now. (though this will not work for `version < v"1.8"` as discussed in the `Compat` issue)

Reply:
Ah, I forgot that `NDTensors` has its own `NDTensors.ndims`. Is that still needed? I forget what that was needed for, as opposed to just using `Base.ndims`. Maybe it was just to work around this issue: https://github.com/ITensor/ITensors.jl/blob/v0.3.57/NDTensors/src/abstractarray/ndims.jl#L8-L10?

If it is still needed, it seems like it make sense to move `NDTensors.ndims` to `TypeParameterAccessors`.

I don't really understand what you are proposing, when you say "should I reroute this function to use `TypeParameters` or should I migrate the code to use `Base.ndims`", which function are you referring to when you say "this function"?

Reply:
@kmp5VT what's the status of this?